### ISSUE TITLE
Issue when providing the `prefix` option after initialization

### ISSUE THREAD

**Issue by zhrichard** [cyrilletuzi/angular-async-local-storage#73]
I tried both v6 and v7 in my app, it works great under chrome and safari, but doesn't work on latest FF.

error stack:
ERROR Error: "Uncaught (in promise): NotFoundError: The operation failed because the requested database object could not be found. For example, an object store did not exist but was being opened.
./node_modules/@ngx-pwa/local-storage/fesm5/ngx-pwa-local-storage.js/IndexedDBDatabase.prototype.transaction/<@http://localhost:4200/vendor.js:82211:24
./node_modules/rxjs/_esm5/internal/operators/map.js/MapSubscriber.prototype._next@http://localhost:4200/vendor.js:93063:22
./node_modules/rxjs/_esm5/internal/Subscriber.js/Subscriber.prototype.next@http://localhost:4200/vendor.js:87777:13
./node_modules/rxjs/_esm5/internal/Subject.js/Subject.prototype.next@http://localhost:4200/vendor.js:87542:25
./node_modules/rxjs/_esm5/internal/ReplaySubject.js/ReplaySubject.prototype.nextInfiniteTimeWindow@http://localhost:4200/vendor.js:87361:9
./node_modules/@ngx-pwa/local-storage/fesm5/ngx-pwa-local-storage.js/IndexedDBDatabase.prototype.connect/<@http://localhost:4200/vendor.js:82182:13
./node_modules/rxjs/_esm5/internal/Subscriber.js/SafeSubscriber.prototype.__tryOrUnsub@http://localhost:4200/vendor.js:87919:13
./node_modules/rxjs/_esm5/internal/Subscriber.js/SafeSubscriber.prototype.next@http://localhost:4200/vendor.js:87857:17
./node_modules/rxjs/_esm5/internal/Subscriber.js/Subscriber.prototype._next@http://localhost:4200/vendor.js:87800:9
./node_modules/rxjs/_esm5/internal/Subscriber.js/Subscriber.prototype.next@http://localhost:4200/vendor.js:87777:13
./node_modules/rxjs/_esm5/internal/operators/tap.js/TapSubscriber.prototype._next@http://localhost:4200/vendor.js:95787:9
./node_modules/rxjs/_esm5/internal/Subscriber.js/Subscriber.prototype.next@http://localhost:4200/vendor.js:87777:13
./node_modules/rxjs/_esm5/internal/operators/take.js/TakeSubscriber.prototype._next@http://localhost:4200/vendor.js:95509:13
./node_modules/rxjs/_esm5/internal/Subscriber.js/Subscriber.prototype.next@http://localhost:4200/vendor.js:87777:13
./node_modules/rxjs/_esm5/internal/observable/race.js/RaceSubscriber.prototype.notifyNext@http://localhost:4200/vendor.js:89868:9
./node_modules/rxjs/_esm5/internal/InnerSubscriber.js/InnerSubscriber.prototype._next@http://localhost:4200/vendor.js:86996:9
./node_modules/rxjs/_esm5/internal/Subscriber.js/Subscriber.prototype.next@http://localhost:4200/vendor.js:87777:13
handler@http://localhost:4200/vendor.js:89093:17
./node_modules/zone.js/dist/zone.js/</ZoneDelegate.prototype.invokeTask@http://localhost:4200/polyfills.js:7463:17
onInvokeTask@http://localhost:4200/vendor.js:48683:24
./node_modules/zone.js/dist/zone.js/</ZoneDelegate.prototype.invokeTask@http://localhost:4200/polyfills.js:7462:17
./node_modules/zone.js/dist/zone.js/</Zone.prototype.runTask@http://localhost:4200/polyfills.js:7230:28
./node_modules/zone.js/dist/zone.js/</ZoneTask.invokeTask@http://localhost:4200/polyfills.js:7538:24
invokeTask@http://localhost:4200/polyfills.js:8582:9
globalZoneAwareCallback@http://localhost:4200/polyfills.js:8608:17

When I went deep into the code, I found somehow the objectStore was not created, seems like 'upgradeneeded' was not be triggered.
If I put a breakpoint in fesm5/ngx-pwa-local-storage.js (line 530), and cleared IndexedDB database in browser dev tool, code would fall there and objectStore would be created, then everything worked fine. But if I just run without breakpoint, the objectStore 'localStorage' would never be created.

Don't know what happened, need your help.

Richard

**Comment by cyrilletuzi** [comment]
All the tests are running well in Firefox 64 on both macOS and Windows. What is your exact version of Firefox?

Be aware that some restrictions can be applied on local storage in some environment (enterprise policy restrictions,...). Can you check you can reproduce the issue in a Firefox with default settings?

**Comment by zhrichard** [comment]
Partly I misused this framework, but the issue happens only on FF. This is what happened, I import localStorage in my common module, then import common module into app module, in app component, I start using localStorage in constructor, I believe a racing issue happens in FF, app component constructor tries to access the db before the object store is created. And because the db is created (object store not), there is no chance later that object store can be created. Why this issue only happens in FF? I would guess FF tries to open db, create object store and proceed transaction operation in an async way and time sequence is not perfectly guaranteed.  I just simply removed localStorage usage in app component and moved it to my common module's init component, then everything was working fine.

**Comment by zhrichard** [comment]
Sorry for confusion. I imported LocalStorageProviders into my common module as I had to specify object store prefix.  And injected LocalStorage service into common service and app component.

To simplify:
-----------------------------------------------
CommonService:
import LocalStorage ...;
CommonService.constructor(private localStorage: LocalStorage) { }
getItem(key) {
  return localStorage.getUnsafeItem(key); 
}
------------------------------------------------
AnotherService:
import CommonService ....
AnotherService.constructor(commonService: CommonService){
  commonService.getItem('abc');
}
-------------------------------------------------
CommonModule:
import CommonService ....
import localStorageProviders ...
...
providers: [
localStorageProviders({ prefix: 'xxxx_' }),
]
--------------------------------------------------
AppModule
....
imports: [
...
CommonModule
]
---------------------------------------------------
App Component
import CommonService ...
import LocalStorage ...
AppComponent.constructor(
private commonService: CommonService,
){
commonService.getItem()...
}
-----------------------------------------------------
Router1 Component  (which will be lazy loaded)
Router1Component.constructor(anotherService: AnotherService){
...
}

**Comment by cyrilletuzi** [comment]
Then it's normal: the `prefix` is used in the object store name. And it must be provided at initialization, ie. in `AppModule`. Otherwise, if provided later, the object store may have been created without the prefix, leading to your issue.

So there's nothing to fix on the lib side, but I'll check later to be sure, because even if you wasn't following the procedure, it feel quite strange the issue could arise.

**Comment by cyrilletuzi** [comment]
Note while it shouldn't happen, starting with v8 (tracked in #76), this error (store not existing) will be catched and propagated to your `.subscribe()`.

Also, v8 has done a complete refactor of the lib, including spotting a few edgy cases, and some Firefox specific, so it may solve this issue (even it's a misuse of the lib).

**Comment by cyrilletuzi** [comment]
Hi @zhrichard, can you test if your initial issue is still there with v8 beta? Instructions in #76.

**Comment by cyrilletuzi** [comment]
@zhrichard ?

**Comment by zhrichard** [comment]
Hi Cyrille,

Sorry for the late reply. Currently it's not easy for me to verify the
case. You probably noticed, once this issue happened, I realized that I
misused your plugin, and then I fixed them by refactoring my code, and it
worked fine. Now I don't have that mis-used case available in my code base
any more, to be honest, I am not quite sure if it's worth to verify it :-)

To upgrade to V8, I didn't do it yet, as for upgrading we have to follow
our product upgrading schedule, I cannot decide by myself. As your V7 works
perfectly fine :-), I guess our team might not considering an immediate
upgrade very soon. But in August we will have a major upgrade, before that
I think I will have chance to verify V8, it's on my to do list now, :-)

Thanks
Richard
